File: Machine Learning A-Z/Part 1 - Data Preprocessing/Section 2 -------------------- Part 1 - Data Preprocessing --------------------/data_preprocessing_template_2.R
# importing libraries
#importing data set

dataset = read.csv('Data.csv')
dataset = dataset[,1:3]
#index starts from 1
#is.na--> if value in column age is missing
dataset$Age = ifelse(is.na(dataset$Age),
                     ave(dataset$Age, FUN = function(x) mean(x, na.rm = TRUE)),
                     dataset$Age
                     )
dataset$Salary = ifelse(is.na(dataset$Salary),
                     ave(dataset$Salary, FUN = function(x) mean(x, na.rm = TRUE)),
                     dataset$Salary)

#factor function will transform them into factors, need not transform them into dummy variables
#press f1 for details
#create a vector of the categorical labels
dataset$Country = factor(dataset$Country, 
                         levels = c('France','Spain','Germany'), 
                         labels = c(1,2,3))

dataset$Purchased = factor(dataset$Purchased, 
                         levels = c('Yes','No'), 
                         labels = c(1,0))

#creating test and train set 
#we need to install caTools library
install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Purchased, SplitRatio = 0.8) #this will return either true or false , true means in training, false means in testing
training_set = subset(dataset, split == TRUE)
testing_set = subset(dataset, split == FALSE)

#feature scaling
training_set[,2:3] = scale(training_set[,2:3])
testing_set[,2:3] = scale(testing_set[,2:3])

File: Machine Learning A-Z/Part 1 - Data Preprocessing/Section 2 -------------------- Part 1 - Data Preprocessing --------------------/data_preprocessing_template2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Jul  2 15:30:46 2017

@author: tgupta2
"""
#data preprocessing
#importing the libraries

import numpy as np #contains mathematical tools
import matplotlib.pyplot as plt#used to plot charts
import pandas as pd #import and manage dataset


#importing the dataset
dataset = pd.read_csv('Data.csv')

#index starts at 0 
#need to create matrix of 3 independent variables
#: means all the lines, : columns --> :-1, means all the columns except the last
X = dataset.iloc[:,:-1].values
                
# : means all the rows, 3 means the third
Y = dataset.iloc[:,3].values

from sklearn.preprocessing import Imputer
#create an object
#strategy is the way you need to replace the missing valyes
#axis = 0--> means of columns
#axis =1--> nean of row
imputer = Imputer(missing_values= 'NaN', strategy='mean', axis = 0)
#we need to mention the columns where we want to fit the imputer, X taking all the lines, and 2 and 3 columns
imputer = imputer.fit(X[:,1:3])                
X[:,1:3] = imputer.transform(X[:,1:3])


#this class encodes labels
from sklearn.preprocessing import LabelEncoder
labelEncoder_x = LabelEncoder()
#this takes the first column, this will return the first encoded of first column
X[:,0]= labelEncoder_x.fit_transform(X[:,0])

from sklearn.preprocessing import OneHotEncoder
#column number of the categorial data
onehotencoder = OneHotEncoder(categorical_features=[0])
X=onehotencoder.fit_transform(X).toarray()


labelEncoder_y = LabelEncoder()
Y= labelEncoder_y.fit_transform(Y)


#splitting data into training and testing set
from sklearn.cross_validation import train_test_split
#test size is size of the testing %
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, train_size = 0.8, random_state = 42)

#feature scaling
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test= sc_X.transform(X_test)

